comparison/comparison.md

```diff
@@ -0,0 +1,375 @@
+
+Loading onto CPU:  
+![Throughput CPU](
+  throughput_timings_size-2000_shuffle_librispeech_runs-10_batch-size16.png "Throughput when loading onto CPU"
+)
+
+Loading onto GPU:  
+![Throughput GPU](
+  throughput_timings_size-2000_shuffle_librispeech_runs-10_batch-size16_gpu.png "Throughput when loading onto GPU"
+)
+
+GPU throughput is higher because a different machine was used.
+The plots were created with following script:
+```python
+# On CPU: python throughput.py
+# On GPU: export CUDA_VISIBLE_DEVICES=0; python throughput.py --gpu
+import os
+from functools import partial
+import time
+
+import numpy as np
+import torch
+from torch.utils.data import DataLoader, Sampler
+import lazy_dataset
+import tqdm
+import click
+
+import matplotlib
+matplotlib.use('Agg')
+import matplotlib.pyplot as plt
+
+# https://github.com/fgnt/paderbox
+from paderbox.io.audioread import load_audio
+from paderbox.utils.nested import flatten
+
+# https://github.com/fgnt/padertorch
+import padertorch as pt
+from padertorch.contrib.jensheit.data import Padder
+
+# exclusive to NT group of Paderborn University
+from padercontrib.database.librispeech import LibriSpeech
+
+
+class AudioReadMap:
+
+    def __init__(self, key):
+        self.key = key
+
+    def __call__(self, example):
+        example['audio_data'] = load_audio(flatten(example)[self.key])
+        return example
+
+
+class ShuffleSampler(Sampler):
+
+    def __init__(self, data_source, seed=None):
+        super(Sampler, self).__init__()
+        self.data_source = data_source
+        self.seed = seed
+
+    def __len__(self):
+        return len(self.data_source)
+
+    def __iter__(self):
+        rng = (
+            np.random if self.seed is None else np.random.RandomState(self.seed)
+        )
+        permutation = np.arange(len(self))
+        rng.shuffle(permutation)
+        return iter(permutation)
+
+
+class WrapDataLoaderToGPU:
+
+    def __init__(self, data_loader, key, device=0, non_blocking=False):
+        """
+        Wraps a torch DataLoader to yield CUDA tensors during iteration
+        """
+        self.loader = data_loader
+        self.key = key
+        self.device = device
+        self.non_blocking = non_blocking
+
+    def __iter__(self):
+        for example in self.loader:
+            if not self.non_blocking:
+                example = pt.data.example_to_device(example, device=self.device)
+            else:
+                example[self.key] = example[self.key].to(
+                    device=self.device, non_blocking=True
+                )
+            yield example
+
+
+def _prepare_lazy_dataset(dataset, num_workers=0, to_gpu=False):
+    as_lazy_dataset = (
+        dataset
+            .shuffle(rng=np.random.RandomState(0))
+            .batch(16, drop_last=False)
+            .map(Padder(padding_keys=['audio_data']))
+    )
+    if num_workers:
+        as_lazy_dataset = as_lazy_dataset.prefetch(
+            num_workers, 16  # prefetch one batch
```

Comment:
Can you change this to `num_workers, 2 * num_workers`?

Reply:
Updated in commit https://github.com/fgnt/lazy_dataset/pull/27/commits/2fe45609f1fec46fd23869f478a0648334f37b6d